File: krishi_web_dev/src/lib/marketplaceService.ts
import axios from 'axios'

const API_BASE_URL = '/api/marketplace'

export interface Product {
  id: string
  name: string
  description: string
  price: number
  discount_price?: number
  stock_quantity: number
  min_order_quantity: number
  unit: string
  images: string[]
  specifications: Record<string, unknown>
  is_active: boolean
  is_featured: boolean
  rating_avg: number
  review_count: number
  product_type: 'buyable' | 'rentable'
  rental_price_per_day?: number
  rental_price_per_week?: number
  rental_price_per_month?: number
  min_rental_days?: number
  max_rental_days?: number
  requires_deposit?: boolean
  deposit_amount?: number
  created_at: string
  updated_at: string
  provider_profiles: {
    id: string
    business_name: string
    city: string
    state: string
    rating_avg: number
    verification_status: string
  }
  categories: {
    id: string
    name: string
  }
}

export interface Category {
  id: string
  name: string
  description: string
  parent_id?: string
  image_url?: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface Order {
  id: string
  farmer_id: string
  provider_id: string
  order_number: string
  status: 'pending' | 'confirmed' | 'shipped' | 'delivered' | 'cancelled'
  total_amount: number
  payment_status: 'pending' | 'paid' | 'failed' | 'refunded'
  payment_method?: string
  shipping_address: Record<string, unknown>
  delivery_date?: string
  notes?: string
  created_at: string
  updated_at: string
  provider_profiles: {
    id: string
    business_name: string
    city: string
    state: string
    phone: string
  }
  order_items: Array<{
    id: string
    quantity: number
    unit_price: number
    total_price: number
    products: {
      id: string
      name: string
      images: string[]
      unit: string
    }
  }>
}

export interface Provider {
  id: string
  user_id: string
  business_name: string
  description: string
  business_type: 'individual' | 'company' | 'cooperative'
  address: string
  city: string
  state: string
  pincode: string
  phone: string
  gst_number?: string
  license_number?: string
  rating_avg: number
  total_orders: number
  verification_status: 'pending' | 'verified' | 'rejected'
  created_at: string
  updated_at: string
  products: Product[]
}

export interface WishlistItem {
  id: string
  farmer_id: string
  product_id: string
  created_at: string
  products: Product
}

export interface RentalBooking {
  id: string
  farmer_id: string
  provider_id: string
  product_id: string
  start_date: string
  end_date: string
  total_days: number
  rental_rate_type: 'daily' | 'weekly' | 'monthly'
  rate_per_period: number
  total_amount: number
  deposit_amount: number
  status: 'pending' | 'confirmed' | 'active' | 'completed' | 'cancelled'
  payment_status: 'pending' | 'paid' | 'failed' | 'refunded'
  delivery_address?: Record<string, unknown>
  pickup_address?: Record<string, unknown>
  notes?: string
  created_at: string
  updated_at: string
  provider_profiles: {
    id: string
    business_name: string
    city: string
    state: string
    phone: string
  }
  products: {
    id: string
    name: string
    images: string[]
    product_type: string
  }
}

export interface RentalAvailability {
  id: string
  product_id: string
  date: string
  is_available: boolean
  created_at: string
}

export class MarketplaceService {
  private static instance: MarketplaceService

  private constructor() {}

  public static getInstance(): MarketplaceService {
    if (!MarketplaceService.instance) {
      MarketplaceService.instance = new MarketplaceService()
    }
    return MarketplaceService.instance
  }

  // Products
  async getProducts(params?: {
    category?: string
    search?: string
    minPrice?: number
    maxPrice?: number
    location?: string
    sortBy?: string
    sortOrder?: 'asc' | 'desc'
    page?: number
    limit?: number
  }) {
    try {
      const searchParams = new URLSearchParams()
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            searchParams.append(key, value.toString())
          }
        })
      }

      const response = await axios.get(`${API_BASE_URL}/products?${searchParams}`)
      return response.data
    } catch (error) {
      console.error('Error fetching products:', error)
      throw error
    }
  }

  async getProduct(id: string) {
    try {
      const response = await axios.get(`${API_BASE_URL}/products/${id}`)
      return response.data
    } catch (error) {
      console.error('Error fetching product:', error)
      throw error
    }
  }

  async createProduct(productData: Partial<Product>) {
    try {
      const response = await axios.post(`${API_BASE_URL}/products`, productData)
      return response.data
    } catch (error) {
      console.error('Error creating product:', error)
      throw error
    }
  }

  async updateProduct(id: string, productData: Partial<Product>) {
    try {
      const response = await axios.put(`${API_BASE_URL}/products/${id}`, productData)
      return response.data
    } catch (error) {
      console.error('Error updating product:', error)
      throw error
    }
  }

  async deleteProduct(id: string) {
    try {
      const response = await axios.delete(`${API_BASE_URL}/products/${id}`)
      return response.data
    } catch (error) {
      console.error('Error deleting product:', error)
      throw error
    }
  }

  // Categories
  async getCategories() {
    try {
      const response = await axios.get(`${API_BASE_URL}/categories`)
      return response.data
    } catch (error) {
      console.error('Error fetching categories:', error)
      throw error
    }
  }

  async createCategory(categoryData: Partial<Category>) {
    try {
      const response = await axios.post(`${API_BASE_URL}/categories`, categoryData)
      return response.data
    } catch (error) {
      console.error('Error creating category:', error)
      throw error
    }
  }

  // Orders
  async getOrders(params?: {
    status?: string
    page?: number
    limit?: number
  }) {
    try {
      const searchParams = new URLSearchParams()
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            searchParams.append(key, value.toString())
          }
        })
      }

      const response = await axios.get(`${API_BASE_URL}/orders?${searchParams}`)
      return response.data
    } catch (error) {
      console.error('Error fetching orders:', error)
      throw error
    }
  }

  async createOrder(orderData: {
    provider_id: string
    items: Array<{
      product_id: string
      quantity: number
      unit_price: number
    }>
    shipping_address: Record<string, unknown>
    notes?: string
  }) {
    try {
      const response = await axios.post(`${API_BASE_URL}/orders`, orderData)
      return response.data
    } catch (error) {
      console.error('Error creating order:', error)
      throw error
    }
  }

  // Providers
  async getProviders(params?: {
    location?: string
    verified?: boolean
    page?: number
    limit?: number
  }) {
    try {
      const searchParams = new URLSearchParams()
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            searchParams.append(key, value.toString())
          }
        })
      }

      const response = await axios.get(`${API_BASE_URL}/providers?${searchParams}`)
      return response.data
    } catch (error) {
      console.error('Error fetching providers:', error)
      throw error
    }
  }

  async createProvider(providerData: Partial<Provider>) {
    try {
      const response = await axios.post(`${API_BASE_URL}/providers`, providerData)
      return response.data
    } catch (error) {
      console.error('Error creating provider:', error)
      throw error
    }
  }

  // Wishlist
  async getWishlist() {
    try {
      const response = await axios.get(`${API_BASE_URL}/wishlist`)
      return response.data
    } catch (error) {
      console.error('Error fetching wishlist:', error)
      throw error
    }
  }

  async addToWishlist(productId: string) {
    try {
      const response = await axios.post(`${API_BASE_URL}/wishlist`, {
        product_id: productId
      })
      return response.data
    } catch (error) {
      console.error('Error adding to wishlist:', error)
      throw error
    }
  }

  async removeFromWishlist(productId: string) {
    try {
      const response = await axios.delete(`${API_BASE_URL}/wishlist?product_id=${productId}`)
      return response.data
    } catch (error) {
      console.error('Error removing from wishlist:', error)
      throw error
    }
  }

  // Rental Bookings
  async getRentals(params?: {
    status?: string
    page?: number
    limit?: number
  }) {
    try {
      const searchParams = new URLSearchParams()
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            searchParams.append(key, value.toString())
          }
        })
      }

      const response = await axios.get(`${API_BASE_URL}/rentals?${searchParams}`)
      return response.data
    } catch (error) {
      console.error('Error fetching rentals:', error)
      throw error
    }
  }

  async createRental(rentalData: {
    product_id: string
    start_date: string
    end_date: string
    rental_rate_type: 'daily' | 'weekly' | 'monthly'
    delivery_address?: Record<string, unknown>
    pickup_address?: Record<string, unknown>
    notes?: string
  }) {
    try {
      const response = await axios.post(`${API_BASE_URL}/rentals`, rentalData)
      return response.data
    } catch (error) {
      console.error('Error creating rental:', error)
      throw error
    }
  }

  // Rental Availability
  async getRentalAvailability(params: {
    product_id: string
    start_date?: string
    end_date?: string
  }) {
    try {
      const searchParams = new URLSearchParams()
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          searchParams.append(key, value.toString())
        }
      })

      const response = await axios.get(`${API_BASE_URL}/rentals/availability?${searchParams}`)
      return response.data
    } catch (error) {
      console.error('Error fetching rental availability:', error)
      throw error
    }
  }

  async updateRentalAvailability(availabilityData: {
    product_id: string
    date: string
    is_available: boolean
  }) {
    try {
      const response = await axios.post(`${API_BASE_URL}/rentals/availability`, availabilityData)
      return response.data
    } catch (error) {
      console.error('Error updating rental availability:', error)
      throw error
    }
  }
}

File: krishi_app/lib/services/marketplace_service.dart
import 'package:flutter/foundation.dart';
import '../models/product.dart' as models;
import '../models/drone_service.dart';
import 'supabase_service.dart';
import 'drone_service_api.dart';

class MarketplaceService {
  static final MarketplaceService _instance = MarketplaceService._internal();
  factory MarketplaceService() => _instance;
  MarketplaceService._internal();

  final SupabaseService _supabaseService = SupabaseService();
  final DroneServiceAPI _droneServiceAPI = DroneServiceAPI();

  /// Get all products with optional filters (including drone services)
  Future<List<models.Product>> getProducts({
    String? category,
    String? search,
    double? minPrice,
    double? maxPrice,
    String? location,
    String? sortBy,
    bool ascending = false,
    int? page,
    int? limit,
  }) async {
    try {
      debugPrint(
        '🛒 [MarketplaceService] Fetching products and drone services...',
      );

      // Get regular products
      final productsData = await _supabaseService.getProducts(
        category: category == 'drone-services' ? null : category,
        search: search,
        minPrice: minPrice,
        maxPrice: maxPrice,
        location: location,
        sortBy: sortBy,
        ascending: ascending,
        page: page,
        limit: limit,
      );

      final products = productsData
          .map((data) => models.Product.fromJson(data))
          .toList();

      // If showing all products or specifically drone services, include drone services
      if (category == null ||
          category == 'all' ||
          category == 'drone-services') {
        final droneServicesData = await _supabaseService.getDroneServices(
          serviceType: category == 'drone-services' ? null : null,
          search: search,
          minPrice: minPrice,
          maxPrice: maxPrice,
          location: location,
          sortBy: sortBy,
          ascending: ascending,
          page: page,
          limit: limit,
        );

        // Transform drone services to product format
        final droneProducts = droneServicesData.map((serviceData) {
          return models.Product.fromJson({
            'id': serviceData['id'],
            'name': serviceData['name'],
            'description': serviceData['description'],
            'price': serviceData['price_per_hour'],
            'stock_quantity': 1, // Drone services are always available
            'min_order_quantity': 1,
            'unit': 'hour',
            'images': serviceData['images'] ?? [],
            'specifications': {
              'service_type': serviceData['service_type'],
              'coverage_area': serviceData['coverage_area'],
              'availability': serviceData['availability'],
              'features': serviceData['features'],
              'is_drone_service': true,
              'contact_phone': serviceData['contact_phone'],
              'contact_email': serviceData['contact_email'],
              'location_address': serviceData['location_address'],
              'location_city': serviceData['location_city'],
              'location_state': serviceData['location_state'],
              'location_pincode': serviceData['location_pincode'],
            },
            'is_active': serviceData['is_active'],
            'is_featured': false,
            'rating_avg': 0.0,
            'review_count': 0,
            'product_type': 'rentable',
            'rental_price_per_day':
                (serviceData['price_per_hour'] as num).toDouble() *
                8, // Approximate daily rate
            'created_at': serviceData['created_at'],
            'updated_at': serviceData['updated_at'],
            'provider_profiles': {
              'id':
                  serviceData['user_profile']?['id'] ?? serviceData['user_id'],
              'business_name':
                  serviceData['user_profile']?['name'] ??
                  'Drone Service Provider',
              'city': serviceData['location_city'] ?? 'Unknown',
              'state': serviceData['location_state'] ?? 'Unknown',
              'rating_avg': 0.0,
              'verification_status': 'verified',
            },
            'categories': {'id': 'drone-services', 'name': 'Drone Services'},
          });
        }).toList();

        products.addAll(droneProducts);
      }

      debugPrint(
        '✅ [MarketplaceService] Retrieved ${products.length} total items (products + drone services)',
      );
      return products;
    } catch (e) {
      debugPrint('❌ [MarketplaceService] Error fetching products: $e');
      rethrow;
    }
  }

  /// Get product categories
  Future<List<models.Category>> getCategories() async {
    try {
      debugPrint('📂 [MarketplaceService] Fetching categories...');

      final categoriesData = await _supabaseService.getCategories();
      final categories = categoriesData
          .map((data) => models.Category.fromJson(data))
          .toList();

      // Add Drone Services category
      categories.add(
        models.Category(id: 'drone-services', name: 'Drone Services'),
      );

      debugPrint(
        '✅ [MarketplaceService] Retrieved ${categories.length} categories (including drone services)',
      );
      return categories;
    } catch (e) {
      debugPrint('❌ [MarketplaceService] Error fetching categories: $e');
      rethrow;
    }
  }

  /// Create an order
  Future<Map<String, dynamic>> createOrder({
    required String productId,
    required int quantity,
    required Map<String, dynamic> shippingAddress,
    String? notes,
  }) async {
    try {
      debugPrint('🛒 [MarketplaceService] Creating order...');

      final order = await _supabaseService.createOrder(
        productId: productId,
        quantity: quantity,
        shippingAddress: shippingAddress,
        notes: notes,
      );

      debugPrint('✅ [MarketplaceService] Order created successfully');
      return order;
    } catch (e) {
      debugPrint('❌ [MarketplaceService] Error creating order: $e');
      rethrow;
    }
  }

  /// Search products
  Future<List<models.Product>> searchProducts(String query) async {
    return getProducts(search: query);
  }

  /// Get products by category
  Future<List<models.Product>> getProductsByCategory(String category) async {
    return getProducts(category: category);
  }

  /// Get featured products
  Future<List<models.Product>> getFeaturedProducts() async {
    return getProducts(sortBy: 'rating_avg', ascending: false);
  }

  /// Get products by price range
  Future<List<models.Product>> getProductsByPriceRange({
    required double minPrice,
    required double maxPrice,
  }) async {
    return getProducts(minPrice: minPrice, maxPrice: maxPrice);
  }

  /// Get products by location
  Future<List<models.Product>> getProductsByLocation(String location) async {
    return getProducts(location: location);
  }

  /// Drone Service Operations

  /// Get all drone services with optional filters
  Future<List<DroneService>> getDroneServices({
    String? serviceType,
    String? search,
    double? minPrice,
    double? maxPrice,
    String? location,
    String? sortBy,
    bool ascending = false,
    int? page,
    int? limit,
  }) async {
    try {
      debugPrint('🚁 [MarketplaceService] Fetching drone services...');
      return await _droneServiceAPI.getDroneServices(
        serviceType: serviceType,
        search: search,
        minPrice: minPrice,
        maxPrice: maxPrice,
        location: location,
        sortBy: sortBy,
        ascending: ascending,
        page: page,
        limit: limit,
      );
    } catch (e) {
      debugPrint('❌ [MarketplaceService] Error fetching drone services: $e');
      rethrow;
    }
  }

  /// Create a drone service
  Future<DroneService> createDroneService({
    required String name,
    required String description,
    required String serviceType,
    required double pricePerHour,
    required String coverageArea,
    required String availability,
    required List<String> features,
    required List<String> images,
    required String contactPhone,
    required String contactEmail,
    required String locationAddress,
    required String locationCity,
    required String locationState,
    required String locationPincode,
  }) async {
    try {
      debugPrint('🚁 [MarketplaceService] Creating drone service...');
      return await _droneServiceAPI.createDroneService(
        name: name,
        description: description,
        serviceType: serviceType,
        pricePerHour: pricePerHour,
        coverageArea: coverageArea,
        availability: availability,
        features: features,
        images: images,
        contactPhone: contactPhone,
        contactEmail: contactEmail,
        locationAddress: locationAddress,
        locationCity: locationCity,
        locationState: locationState,
        locationPincode: locationPincode,
      );
    } catch (e) {
      debugPrint('❌ [MarketplaceService] Error creating drone service: $e');
      rethrow;
    }
  }

  /// Search drone services
  Future<List<DroneService>> searchDroneServices(String query) async {
    return _droneServiceAPI.searchDroneServices(query);
  }

  /// Get drone services by service type
  Future<List<DroneService>> getDroneServicesByType(String serviceType) async {
    return _droneServiceAPI.getDroneServicesByType(serviceType);
  }

  /// Get drone services by price range
  Future<List<DroneService>> getDroneServicesByPriceRange({
    required double minPrice,
    required double maxPrice,
  }) async {
    return _droneServiceAPI.getDroneServicesByPriceRange(
      minPrice: minPrice,
      maxPrice: maxPrice,
    );
  }

  /// Get drone services by location
  Future<List<DroneService>> getDroneServicesByLocation(String location) async {
    return _droneServiceAPI.getDroneServicesByLocation(location);
  }

  /// Get featured drone services
  Future<List<DroneService>> getFeaturedDroneServices() async {
    return _droneServiceAPI.getFeaturedDroneServices();
  }

  /// Get available service types
  List<String> getServiceTypes() {
    return _droneServiceAPI.getServiceTypes();
  }

  /// Get available features
  List<String> getAvailableFeatures() {
    return _droneServiceAPI.getAvailableFeatures();
  }
}
